### PR TITLE
start python-based testing framework

### PR DESCRIPTION
**Pull request by jlevon**

### CODE REVIEW

File: test/py/libvfio_user.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+def connect_sock():
+    sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
+    sock.connect(SOCK_PATH)
+    return sock
+
+def get_reply(sock, expect=0):
+    buf = sock.recv(4096)
+    (msg_id, cmd, msg_size, flags, errno) = struct.unpack("HHIII", buf[0:16])
+    assert errno == expect
+    msg_size -= 16
```
Comment (tmakatos) [inline test/py/libvfio_user.py:108]: What's the point of this?

File: test/py/libvfio_user.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+                      0, error_no)
+
+    msg_id += 1
+
+    return buf
+
+def vfu_create_ctx(trans=VFU_TRANS_SOCK, sock_path=SOCK_PATH, flags=0,
+                   private=None, dev_type=VFU_DEV_TYPE_PCI):
+    if os.path.exists(sock_path):
+        os.remove(sock_path)
```
Comment (tmakatos) [inline test/py/libvfio_user.py:139]: nit: just attempt to remove the path and handle the `ENOENT` exception?

File: test/py/libvfio_user.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    return ret
+
+def vfu_run_ctx(ctx):
+    lib.vfu_run_ctx(ctx)
+
+def vfu_destroy_ctx(ctx):
+    lib.vfu_destroy_ctx(ctx)
+    ctx = None
+    if os.path.exists(SOCK_PATH):
+        os.remove(SOCK_PATH)
```
Comment (tmakatos) [inline test/py/libvfio_user.py:167]: ditto

**Review by tmakatos — APPROVED** [review]

File: Makefile [hunk 4]
```diff
@@ -54,13 +54,60 @@ BUILD_DIR = $(BUILD_DIR_BASE)/$(BUILD_TYPE)
...
+PYTESTVALGRIND = \
+	BUILD_TYPE=$(BUILD_TYPE) \
+	PYTHONMALLOC=malloc \
+	valgrind \
+	--suppressions=$(CURDIR)/test/py/valgrind.supp \
+	--quiet \
+	--errors-for-leak-kinds=definite \
+	--show-leak-kinds=definite \
+	--leak-check=full \
+	--error-exitcode=1 \
```
[the remark this answers is not in the document]
Comment (jlevon) [reply Makefile:91]: sure, thanks.

File: test/py/libvfio_user.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+LIBVFIO_USER_MINOR = 1
+
+VFIO_USER_CLIENT_MAX_FDS_LIMIT = 1024
+
+SERVER_MAX_FDS = 8
+
+SERVER_MAX_MSG_SIZE = 65536
+
+# enum vfio_user_command
+VFIO_USER_VERSION	    			= 1
```
[the remark this answers is not in the document]
Comment (jlevon) [reply test/py/libvfio_user.py:59]: done

File: test/py/libvfio_user.py (hunk 1 above)
Comment (jlevon) [reply test/py/libvfio_user.py:108]: none, thanks

File: test/py/libvfio_user.py (hunk 2 above)
Comment (jlevon) [reply test/py/libvfio_user.py:138]: I think it's clearer as is?

Comment (tmakatos) [reply test/py/libvfio_user.py:138]: In theory this is vulnerable to TOCTOU bugs, but we shouldn't have to deal with things like that in the testing framework.

**Pull request merged**